#!/bin/bash
#PBS -l walltime=24:0:0 
#PBS -q uv@wagap-pro.cerit-sc.cz 
#PBS -l select=1:ncpus=1:mem=40gb:scratch_local=80gb
#PBS -j oe
#PBS -N 03v3_RSEM_index
#############################################################################################
# Prepare RSEM alignment index
#
# Requires RSEM, pigz, bowtie or alernatively STAR aligner
#
# --star option will allow RSEM to use the STAR aligner
# To use an alternative alignment program, align the input reads against the file reference_name.idx.fa generated by rsem-prepare-reference, 
# and format the alignment output in SAM/BAM/CRAM format. Then, instead of providing reads to rsem-calculate-expression, 
# specify the --alignments option and provide the SAM/BAM/CRAM file as an argument.
# Prepare RSEM alignment index
#
# Requires RSEM, pigz, bowtie or alernatively STAR aligner
#############################################################################################
## initialize the required application
# module add rsem-1.2.8 # Metacentrum has only an old version that doesn't support STAR aligner
module add star-2.5.2b
# module add bowtie-1.0.0
module add bowtie2-2.3.0
############################################################################################
### Variables
DATADIR="/storage/brno3-cerit/home/tskalicky/genomes/human/ensembl91"
GENOME="/storage/brno3-cerit/home/tskalicky/genomes/human/ensembl91/GRCh38.91.dna.primary_assembly.fa.gz"
GTF="/storage/brno3-cerit/home/tskalicky/genomes/human/ensembl91/GRCh38.91.gtf.gz"
GTF_MODIF="/storage/brno3-cerit/home/tskalicky/genomes/human/ensembl91/GRCh38.91_with_tRNA_modif_for_RSEM.sorted.gtf"
INDEX_OUTPUT_DIR="/storage/brno3-cerit/home/tskalicky/genomes/human/ensembl91/RSEM_index/star_align"
#
THREADS=$PBS_NUM_PPN
#
# Adding RSEM binaries into the PATH
export PATH="/storage/brno3-cerit/home/tskalicky/tools/RSEM-1.3.0/bin:$PATH"
#
# Binaries
RSEM_BIN="/storage/brno3-cerit/home/tskalicky/tools/RSEM-1.3.0/bin"
STAR_BIN="$(which STAR)"

############################################################################################
# copy input data using SCRATCHDIR storage which is shared via NFSv4
# clean the SCRATCH when job finishes (and data
# are successfully copied out) or is killed
# use cp -avr when copying directories
cd $DATADIR
trap 'clean_scratch' TERM EXIT # sets up scratch cleaning in case an error occurs
cp -av $GENOME $GTF $GTF_MODIF $SCRATCHDIR
cd $SCRATCHDIR

if [ ! -d "$SCRATCHDIR" ] ; then echo "Scratch directory is not created!" 1>&2; exit 1; fi #checks if scratch directory is created
echo "SCRATCHDIR path is:" $SCRATCHDIR

####################################################################################################
### Genome and annotation preparation
# "basename" will print file NAME with any leading directory components removed
# "dirname" will print DIRECTORY PATH without filename
GENOME_NAME=$(basename $GENOME)
GTF_NAME=$(basename $GTF)

unpigz -p $THREADS $GENOME_NAME
unpigz -p $THREADS $GTF_NAME

GENOME2=${GENOME_NAME%.*}
GTF2=${GTF_NAME%.*}
GEN_DIR=${GENOME_NAME%.fa*}
# ENSEMBL changed the gtf format and RSEM has problem with missing "transcript_id" patern. We need to filter the *.gtf file using awk for example

#awk '$3 == "exon"' $GTF_NAME > $GTF2_modif.gtf

#GTF3=$GTF2_modif.gtf

############################################################################################
# Read length for sjdbOverhang; --sjdbOverhang 100 should work fine for most of the data, but more specific setting based on the real read length should be more sensitive https://groups.google.com/forum/#!msg/rna-star/h9oh10UlvhI/BfSPGivUHmsJ
RD_LENGTH=79 # Read length from the sequencing. Illumina sometimes reports N+1 http://seqanswers.com/forums/archive/index.php/t-31154.html; in case you change this value uncomment next line as well
#RD_LENGTH=$[$RD_LENGTH-$TRIM_LEFT-$TRIM_RIGHT-1] # Modified read length for the index creation and mapping - should be read length -1
#RD_LENGTH changed to 79 because Dis3L2 PE libs after trimminf are in range of 17-80 bases
# ############################################################################################
### Indexing - RUN ONLY ONCE FOR ONE GENOME AND RSEM VERSION 
mkdir $SCRATCHDIR/rsem_STAR_index
#rsem-prepare-reference --gtf $SCRATCHDIR/$GTF3 --bowtie --bowtie2 $GENOME2 $SCRATCHDIR/rsem_index/$GEN_DIR
rsem-prepare-reference --gtf $SCRATCHDIR/GRCh38.91_with_tRNA_modif_for_RSEM.sorted.gtf \
--star --star-path $STAR_BIN \
--star-sjdboverhang $RD_LENGTH $GENOME2 $SCRATCHDIR/rsem_STAR_index/$GEN_DIR
############################################################################################
### Copy data from scratch back to home dir and clean scratch
mkdir -p $INDEX_OUTPUT_DIR
rm $GENOME2 $GTF2
cp -avr $SCRATCHDIR/rsem_STAR_index/ $INDEX_OUTPUT_DIR/ || export CLEAN_SCRATCH=false
# rm -r $SCRATCHDIR/*
echo "Script finished on:"
date +"%d/%m/%Y %H:%M:%S $HOSTNAME"


